js/reaction.js

```diff
@@ -651,16 +651,53 @@ function removeSlowly(button, pattern, callback) {
     buttonsToClick =
         buttonsToClick.add($(this).children('legend').find('.validate_button'));
   });
+  makeUndoable(node);
   node.hide('slow', function() {
-    node.remove();
     buttonsToClick.trigger('click');
-    if (callback !== undefined) {
-      callback();
-    }
+    ord.inputs.updateSidebar();
+  });
+  dirty();
+}
+
+/**
+ * Reverses the hide() in the most recent invocation of removeSlowly().
+ * Removes the node's "undo" button. Does not trigger validation.
+ */
+function undoSlowly() {
+  $('.undoable').removeClass('undoable').show('slow');
+  $('.undo').not('#undo_template').hide('slow', function() {
+    $(this).remove();
+    ord.inputs.updateSidebar();
   });
   dirty();
 }
 
+/**
+ * Marks the given node for possible future undo. Adds an "undo" button to do
+ * it. Deletes any preexisting undoable nodes and undo buttons.
+ * @param {!Node} node The DOM fragment to hide and re-show.
+ */
+function makeUndoable(node) {
+  $('.undoable').remove();
+  node.addClass('undoable');
+  $('.undo').not('#undo_template').remove();
+  const button = $('#undo_template').clone();
+  button.removeAttr('id');
+  button.hide();
```

Comment:
What's the purpose of the hide() and then show()?

Reply:
The hide() is redundant. Thanks!